### PR TITLE
[mono] Transition memory allocations from image sets to memory managers.

### PR DESCRIPTION
**Pull request by vargaz**
* Move hashes from MonoImageSet to MonoGenericMemoryManager.
* Add a mono_mem_manager_get_generic () function which returns
  a memory manager for a set images. Same as the get_image_set ()
  function.
* Transition users of image sets to use MonoGenericMemoryManager.

### CODE REVIEW

**Comment by vargaz** [comment]
Issues:
- the image set code still needs to be removed
- Having a separate MonoGenericMemoryManager type is very awkward. Requires a lot
  of casts to call normal memory manager functions on it, or a separate variable.
  Also, there are now 2 memory managers per alc, the 'normal' one, and the 'generic' one which
  handles generic instances which belong only to that one 'alc'.
  It might be better to just merge MonoGenericMemoryManager into the normal one.

File: src/mono/mono/metadata/class-internals.h [hunk 1]
```diff
@@ -405,7 +405,7 @@ struct _MonoMethodInflated {
 	} method;
 	MonoMethod *declaring;		/* the generic method definition. */
 	MonoGenericContext context;	/* The current instantiation */
-	MonoImageSet *owner; /* The image set that the inflated method belongs to. */
+	MonoGenericMemoryManager *owner; /* The mem manager that the inflated method belongs to. */
```
Comment (lambdageek) [inline src/mono/mono/metadata/class-internals.h:408]: I think this should be a `MonoMemoryManager *owner;` and you should downcast to `MonoGenericMemoryManager` in the memory manager utility functions.  Treat `MonoMemoryManager` like an abstract base class.

In the places where you need to do stuff with its ALCs, downcast in utility functions that worth with either a Singleton memory manager or a Generic one.

Comment (vargaz) [reply src/mono/mono/metadata/class-internals.h:408]: Why is that better ? I think the best approach would be getting rid of these subclasses altogether, memory managers are not very numerous, so it doesn't save too much space, just complicates code and makes it more error prone. Subclasses are not very well supported by C.

Comment (lambdageek) [reply src/mono/mono/metadata/class-internals.h:408]: yea, ok.

@CoffeeFlux do we lose anything in collectability if we combine generic and singleton mem managers into one struct?

Comment (vargaz) [reply src/mono/mono/metadata/class-internals.h:408]: So the mem manager functions take a MonoMemoryManager, but the caches are in the generic version, so no matter what the type of this field is, you need either casts or 2 variables.

Comment (lambdageek) [reply src/mono/mono/metadata/class-internals.h:408]: I think using a single memory manager struct is ok.

If we just have a single memory manager struct, do you think the caches should be in it?  I think you use memory managers in the JIT in some places - and I guess we will have to say that the caches in those have to be NULL.

Is it possible to confuse one memory manager for another and write something to a cache accidentally?

Maybe we at least need a kind enum on the mem manager struct that we can assert if it's a JIT mm the caches are null when we destroy the mm?

Comment (vargaz) [reply src/mono/mono/metadata/class-internals.h:408]: The JIT memory manager is an extension of a normal memory manager, accessed through a pointer. So the same struct owns both metadata and JIT data.

File: src/mono/mono/metadata/class-init.c [hunk 2]
```diff
@@ -1004,22 +1003,25 @@ mono_class_create_bounded_array (MonoClass *eclass, guint32 rank, gboolean bound
...
 	cached = NULL;
 	if (rank == 1 && !bounded) {
-		if (image_set) {
-			mono_image_set_lock (image_set);
-			cached = (MonoClass *)g_hash_table_lookup (image_set->szarray_cache, eclass);
-			mono_image_set_unlock (image_set);
+		if (mm) {
+			mono_mem_manager_lock (mm);
+			if (!mm->szarray_cache)
+				mm->szarray_cache = g_hash_table_new_full (mono_aligned_addr_hash, NULL, NULL, NULL);
```
Comment (CoffeeFlux) [inline src/mono/mono/metadata/class-init.c:1022]: Do these lazily-created hash tables ever get freed? Am I just missing it?

**Review by CoffeeFlux — APPROVED** [review]
Sorry for the delay on this.

The general idea of merging the memory manager types seems fine to me. Only concern is that the new members are all getting freed, which I don't see for some of them.

File: src/mono/mono/metadata/class-init.c (hunk 2 above)
Comment (vargaz) [reply src/mono/mono/metadata/class-init.c:1022]: Not right now, but right now memory managers are never freed.

Comment (CoffeeFlux) [reply src/mono/mono/metadata/class-init.c:1022]: We do have a free function though - it would be nice to add those ASAP before we forget.

**Pull request merged**